### PR TITLE
Provide AddImport/AddExport for each element in the C-API

### PR DESCRIPTION
**Pull request by dcodeIO**
As discussed in #1274, this PR updates the C-API to include proper use of `ExternalKind` (probably introduced after the C-API was created) for each import/export element.

Introduces the following functions:

* **BinaryenAddFunctionImport**
* **BinaryenAddTableImport**
* **BinaryenAddMemoryImport**
* **BinaryenAddGlobalImport**
* **BinaryenAddFunctionExport**
* **BinaryenAddTableExport**
* **BinaryenAddMemoryExport**
* **BinaryenAddGlobalExport** (though not supported [if mutable](https://github.com/WebAssembly/binaryen/pull/1292#issuecomment-344680113))

Deprecates the following functions:

* **BinaryenAddImport** mapped to **BinaryenAddFunctionImport**
* **BinaryenAddExport** mapped to **BinaryenAddFunctionExport**

Also updates binaryen.js accordingly.

### CODE REVIEW

**Comment by binji** [comment]
> BinaryenAddGlobalExport (though not supported in the MVP)

Global imports and exports are supported in the MVP as long as they're immutable.

File: src/binaryen-c.h [hunk 1]
```diff
@@ -402,14 +413,22 @@ BinaryenFunctionRef BinaryenAddFunction(BinaryenModuleRef module, const char* na
...
+BinaryenImportRef BinaryenAddMemoryImport(BinaryenModuleRef module, const char* internalName, const char* externalModuleName, const char *externalBaseName);
+BinaryenImportRef BinaryenAddGlobalImport(BinaryenModuleRef module, const char* internalName, const char* externalModuleName, const char *externalBaseName, BinaryenType globalType);
 void BinaryenRemoveImport(BinaryenModuleRef module, const char* internalName);
 
 // Exports
 
 typedef void* BinaryenExportRef;
 
-BinaryenExportRef BinaryenAddExport(BinaryenModuleRef module, const char* internalName, const char* externalName);
+WASM_DEPRECATED BinaryenExportRef BinaryenAddExport(BinaryenModuleRef module, const char* internalName, const char* externalName);
```
[the remark this answers is not in the document]
Comment (dcodeIO) [reply src/binaryen-c.h]: I believe no matter how we do it, the API will remain unnecessarily complex unless (here) imports and exports inherit their function or global type from the respective internal element. I know that they don't for a good reason, that is everything remains independent of each other, but API-wise...

[the remark this answers is not in the document]
Comment (kripken) [reply src/binaryen-c.h]: I don't understand, can you elaborate?

[the remark this answers is not in the document]
Comment (dcodeIO) [reply src/binaryen-c.h]: That is, for example, having `BinaryenAddFunctionImport` for function imports and `BinaryenAddExport(..., BinaryenExternalKind)` for function exports - or otherwise not-so-useful but symmetric functions. All this boils down to imports requiring a kind, a variable type for globals respectively a function type for functions or nothing at all for everything else. If that wouldn't be necessary (here; instead inherited from the linked internal element), we could just do a uniform AddImport/AddExport without worrying about these details in the API.

[the remark this answers is not in the document]
Comment (kripken) [reply src/binaryen-c.h]: What do you mean by "inherited from the linked internal element"? We can inspect the internal element as necessary in the C API implementation code, do you mean something more than that?

Sorry, I feel like I'm not following your line of thought here. What is this arguing for?

[the remark this answers is not in the document]
Comment (dcodeIO) [reply src/binaryen-c.h]: Sorry, just thinking about how to do it.

1. Current state of this PR: One `Add*Import` for each element and one `Add*Export` for each element, with some import types taking specific parameters (i.e. functionType, globalType). No `ExternalKind`s to specify anywhere.

2. One `Add*Import` for each element with some imports taking specific parameters (i.e. functionType for `Function` imports, type for `Global` imports), but just one `AddExport` taking an `ExternalKind`, making parameters somewhat asymmetric. If I understand correctly, that corresponds to your initial comment.

3. If some `Import`s wouldn't need `functionType` or `type`, we could just make a single `AddImport` taking an `ExternalKind`, just like for `AddExport`. Not sure if that can be implemented within the internal API or if it could otherwise be emulated in just the C-API, but that's basically what I tried and obviously failed to elaborate on.

TL;DR Not sure what's best either.

[the remark this answers is not in the document]
Comment (kripken) [reply src/binaryen-c.h]: I see now, thanks. 2 is what I was thinking of, but I forgot you'd need the extra `kind` argument. About 3, so far all exports need `functionType` (functions) or `type` (globals), and not likely to change any time soon.

Overall I think this PR (option 1) seems reasonable to go with. Please just fix the conflict before we merge.

**Comment by dcodeIO** [comment]
Just noticed: The C-API assumes that there is at least one memory, using name "0". Hence, `Module#addMemoryImport` works with "0" as the name only. Even when renaming utilities would be in place, the memory would have to be manually renamed from "0" to something else first.

Possible not-so-nice solutions:

* Just document, for now
* Accept an additional `name` parameter in `Module#setMemory`, breaking the API
* Deprecate `Module#setMemory` in favor of a new `Module#addMemory` once multiple memories are supported, but not sure if that's even on the roadmap (the same goes for tables)

**Comment by kripken** [comment]
Multiple memories are pretty far out, there's no roadmap for them that I am aware of. I'd say we can just document this as a current limitation.

**Comment by binji** [comment]
Yeah, there are no proposals currently. Though they've come up a number of times at the CG recently. I personally think they'd help with a current limitation of WebAssembly, where we require a copy-in/copy-out for any memory buffers that live outside linear memory (such as video or audio buffers).

I think the biggest issue is that having more than one memory is difficult for producers (C/C++ don't really make this easy, though there is some limited support for segmented memory AIUI), and for consumers (many engines pin the memory base pointer for quick lookup).
